pkg/lib/ratelimit/gcra_test.go

```diff
@@ -0,0 +1,172 @@
+				{time: "10s", n: 1, ok: true, timeToAct: 15},
+
+				{time: "10s", n: 1, ok: false, timeToAct: 15},
+				{time: "11s", n: 0, ok: true, timeToAct: 15},
+				{time: "11s", n: 1, ok: false, timeToAct: 15},
+
+				{time: "100s", n: 0, ok: true, timeToAct: 85},
+				{time: "100s", n: 1, ok: true, timeToAct: 90},
+				{time: "100s", n: 1, ok: true, timeToAct: 95},
+				{time: "100s", n: 1, ok: true, timeToAct: 100},
+				{time: "100s", n: 1, ok: true, timeToAct: 105},
+				{time: "100s", n: 1, ok: false, timeToAct: 105},
+				{time: "100s", n: 0, ok: true, timeToAct: 105},
+			},
+		})
+		test("substained requests", &schedule{
+			period: "20s",
+			burst:  4,
+			entries: []scheduleEntry{
+				{time: "0s", n: 1, ok: true, timeToAct: -10},
+				{time: "5s", n: 1, ok: true, timeToAct: -5},
+				{time: "10s", n: 1, ok: true, timeToAct: 0},
+				{time: "15s", n: 1, ok: true, timeToAct: 5},
+				{time: "20s", n: 1, ok: true, timeToAct: 10},
+				{time: "25s", n: 1, ok: true, timeToAct: 15},
+				{time: "30s", n: 1, ok: true, timeToAct: 20},
+			},
+		})
+		test("burst requests", &schedule{
+			period: "20s",
+			burst:  4,
+			entries: []scheduleEntry{
+				{time: "0s", n: 5, ok: false, timeToAct: -15},
+				{time: "0s", n: 4, ok: true, timeToAct: 5},
+				{time: "1s", n: 1, ok: false, timeToAct: 5},
+				{time: "1s", n: 0, ok: true, timeToAct: 5},
+
+				{time: "100s", n: 5, ok: false, timeToAct: 85},
+				{time: "100s", n: 3, ok: true, timeToAct: 100},
+				{time: "101s", n: 1, ok: true, timeToAct: 105},
+				{time: "101s", n: 0, ok: true, timeToAct: 105},
+
+				{time: "200s", n: 5, ok: false, timeToAct: 185},
+				{time: "200s", n: 2, ok: true, timeToAct: 195},
+				{time: "201s", n: 2, ok: true, timeToAct: 205},
+				{time: "201s", n: 0, ok: true, timeToAct: 205},
+			},
+		})
+		test("reservation & return", &schedule{
+			period: "10s",
+			burst:  2,
+			entries: []scheduleEntry{
+				{time: "0s", n: 0, ok: true, timeToAct: -5},
+				{time: "0s", n: 1, ok: true, timeToAct: 0},
+				{time: "0s", n: -1, ok: true, timeToAct: -5},
+
+				{time: "0s", n: 1, ok: true, timeToAct: 0},
+				{time: "3s", n: 1, ok: true, timeToAct: 5},
+				{time: "6s", n: -1, ok: true, timeToAct: 0},
+
+				{time: "8s", n: 1, ok: true, timeToAct: 8},
+				{time: "9s", n: 1, ok: true, timeToAct: 13},
+				{time: "10s", n: -1, ok: true, timeToAct: 8},
+				{time: "11s", n: -1, ok: true, timeToAct: 3},
+				{time: "20s", n: -1, ok: true, timeToAct: 10},
+				{time: "21s", n: 1, ok: true, timeToAct: 21},
+				{time: "22s", n: 1, ok: true, timeToAct: 26},
+				{time: "23s", n: 1, ok: false, timeToAct: 26},
+			},
+		})
+		test("ignore old rate limit state", &schedule{
+			period: "10s",
+			burst:  2,
+			entries: []scheduleEntry{
+				{fn: func() {
+					s.HSet(testKey, "test", "123")
+				}},
+				{time: "0s", n: 0, ok: true, timeToAct: -5},
+				{time: "0s", n: 1, ok: true, timeToAct: 0},
+				{time: "0s", n: -1, ok: true, timeToAct: -5},
+			},
+		})
```

Comment:
Does the new algorithm handle negative underflow now? Please add a test case to demonstrate it.

Reply:
It is tested in the reservation test case, let me add some comments to describe the entries.